components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/FileEncryptionUtility.java

```diff
@@ -0,0 +1,245 @@
+import java.util.List;
+import javax.crypto.Cipher;
+import javax.crypto.CipherInputStream;
+import javax.crypto.CipherOutputStream;
+import javax.crypto.KeyGenerator;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.spec.SecretKeySpec;
+
+import static org.wso2.carbon.apimgt.core.util.APIMgtConstants.EncryptionConstants;
+
+/**
+ *  Class for File Encryption Related functions
+ */
+public class FileEncryptionUtility {
+    private static final Logger log = LoggerFactory.getLogger(FileEncryptionUtility.class);
+    private static final FileEncryptionUtility instance = new FileEncryptionUtility();
+    public static final String SECURITY_DIR = File.separator + "resources" + File.separator + "security";
+    private static final int AES_Key_Size = 128;
+
+    private FileEncryptionConfigurations config;
+    private String aesKeyFileLocation;
+    private SecureVault secureVault;
+
+    /**
+     * Gives FileEncryptionUtility instance
+     *
+     * @return FileEncryptionUtility object
+     */
+    public static FileEncryptionUtility getInstance() {
+        return instance;
+    }
+
+    /**
+     * Sets the location to store the encrypted AES key file
+     * Sets secure vault instance to encrypt AES key
+     * Calls createAndStoreAESKey() method
+     *
+     * @throws APIManagementException if an error occurs while initializing the file encryption
+     */
+    public void init() throws APIManagementException {
+        setConfig(ServiceReferenceHolder.getInstance().getAPIMConfiguration().getFileEncryptionConfigurations());
+        setAesKeyFileLocation();
+        setSecureVault(ServiceReferenceHolder.getInstance().getSecureVault());
+        if (secureVault == null) {
+            throw new APIManagementException("Secure vault OSGi service cannot be accessed");
+        }
+        createAndStoreAESKey();
+    }
+
+    /**
+     * Encrypts the contents of a file and stores it in a new file
+     *
+     * @param inputFilePath    absolute path of the file to encrypt
+     * @param outputFilePath   expected absolute path of the new encrypted file
+     * @throws APIManagementException  if an error occurs encrypting the file
+     */
+    public void encryptFile(String inputFilePath, String outputFilePath) throws APIManagementException {
+        try {
+            Cipher aesCipher = Cipher.getInstance(EncryptionConstants.AES);
+            SecretKeySpec aesKeySpec = new SecretKeySpec(getAESKey(), EncryptionConstants.AES);
+            aesCipher.init(Cipher.ENCRYPT_MODE, aesKeySpec);
+
+            Files.deleteIfExists(Paths.get(outputFilePath));
+            InputStream inputStream = APIFileUtils.readFileContentAsStream(inputFilePath);
+            CipherOutputStream cipherOutStream = new CipherOutputStream(
+                    new FileOutputStream(outputFilePath), aesCipher);
+            IOUtils.copy(inputStream, cipherOutStream);
+            inputStream.close();
+            cipherOutStream.close();
```

Comment:
This is a suggestion to use the "try with resources"[1] for the inputStream and cipherOutputStream. Then you will not have to close the streams explicitly.

[1] https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

Reply:
I went through the material. Thanks for the suggestion. "try" with resources worked well with  inputStream and cipherOutputStream. Then I also check the code-coverage by the tests, but that had gone down to 40% from 83% (branch coverage). Shall we leave this part as it is?

Reply:
I don't see an issue leaving it like this. Apparently there is some issue with unit test coverage when using try with resources. But there seems to be a work-around.
https://stackoverflow.com/questions/26360245/try-with-resource-unit-test-coverage

Reply:
Thanks. I'll check that too.

Reply:
@SuKSW shall we keep close() inside a finally block at least and keep stream creation code inside the same try block? otherwise this will lead to resource leakages